### PR TITLE
message queue implementation in the pComponent

### PR DESCRIPTION
**Pull request by LiuHaolan**

### CODE REVIEW

File: include/lego/rpc/struct_p2m.h [hunk 1]
```diff
@@ -194,6 +194,62 @@ struct m2p_execve_struct {
...
+/*
+void handle_mq_close_request(struct p2m_mqclose_payload* payload, 
+	struct thpool_buffer*tb);
+*/
+
+
+/* for now we use fixed payload here
+ * we can make run-time decision about msg_size in the future
+ */
+#define MAX_MSG_LENGTH 100
```
Comment (yiying-zhang) [inline include/lego/rpc/struct_p2m.h]: Can you make MAX_MSG_LENGTH a config in kconfig?

File: include/lego/rpc/struct_p2m.h [hunk 2]
```diff
@@ -194,6 +194,62 @@ struct m2p_execve_struct {
...
+	char msg[MAX_MSG_LENGTH];
+	char mq_name[MAX_FILENAME_LENGTH];
+};
+
+/*
+void handle_mq_send_request(struct p2m_mqsend_payload* payload, 
+	struct thpool_buffer*tb);
+*/
+
+#define MAX_MSG_LENGTH 100
```
Comment (yiying-zhang) [inline include/lego/rpc/struct_p2m.h]: Remove this. It's redundant.

File: managers/memory/missing_syscalls.c [hunk 3]
```diff
@@ -284,3 +284,23 @@ SYSCALL_DEFINE1(fsync, unsigned int, fd)
...
+SYSCALL_DEFINE2(mq_close, char*, name, unsigned long, name_size)
+{
+        BUG();
+}
+
+SYSCALL_DEFINE3(mq_open, char*, name, unsigned long, name_size, unsigned long, msg_size)
+{
+	BUG();
+}
+
```
Comment (yiying-zhang) [inline managers/memory/missing_syscalls.c:306]: These are not implemented??

File: managers/processor/message_queue.c [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+#include "processor.h"
+#include <lego/fit_ibapi.h>
+
+SYSCALL_DEFINE4(mq_send, char*, name, unsigned long, name_size, unsigned long, msg_size, const char*, msg_data)
+{
+	ssize_t retval, retlen;
+	u32 len_msg;
+	void *msg;
+	struct common_header* hdr;
+	struct p2m_mqsend_payload* payload;
```
Comment (yiying-zhang) [inline managers/processor/message_queue.c:29]: add an empty line below all the variable declaration

File: managers/processor/message_queue.c [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	len_msg = sizeof(*hdr)+sizeof(*payload);
+	msg = kmalloc(len_msg, GFP_KERNEL);
+	if(!msg)
+		return -ENOMEM;
+
+	hdr = (struct common_header *)msg;
+	hdr->opcode = P2M_MQSEND;
+	hdr->src_nid = LEGO_LOCAL_NID;
+	payload=to_payload(msg);
```
Comment (yiying-zhang) [inline managers/processor/message_queue.c]: formatting. add space before and after =

File: managers/processor/message_queue.c [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+ 
+	copy_from_user(payload->mq_name, name, name_size+1);
+	copy_from_user(payload->msg, msg_data, msg_size+1);
+	payload->msg_size=msg_size;
```
Comment (yiying-zhang) [inline managers/processor/message_queue.c]: formatting

File: managers/processor/message_queue.c [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+	retlen = ibapi_send_reply_imm(current_pgcache_home_node(), msg, len_msg, &retval, sizeof(retval),false);
```
Comment (yiying-zhang) [inline managers/processor/message_queue.c:45]: formatting, space before false

File: managers/processor/message_queue.c [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+ 
+	copy_from_user(payload->mq_name, name, name_size+1);
+	copy_from_user(payload->msg, msg_data, msg_size+1);
+	payload->msg_size=msg_size;
+
+	retlen = ibapi_send_reply_imm(current_pgcache_home_node(), msg, len_msg, &retval, sizeof(retval),false);	
+
+	/* check return value */
+	if(retlen == -ETIMEDOUT){
+		return -1;
```
Comment (yiying-zhang) [inline managers/processor/message_queue.c:50]: still need to free msg even if return -1

File: managers/processor/message_queue.c [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+	copy_from_user(payload->mq_name, name, name_size+1);
+
+	retlen = ibapi_send_reply_imm(current_pgcache_home_node(), msg, len_msg, &retval, sizeof(retval),false);	
+
+	/* check return value */
+	if(retlen == -ETIMEDOUT){
+		return -1;
+	}		
+
+
```
Comment (yiying-zhang) [inline managers/processor/message_queue.c:89]: too many empty lines

File: managers/processor/message_queue.c [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+SYSCALL_DEFINE4(mq_receive, char*, name, unsigned long, name_size, unsigned long*, msg_size, char*, msg_data)
+{	
+	struct p2m_mqrecv_reply_struct retval; 
+	ssize_t retlen;
+	u32 len_msg;
+	void *msg;
+	struct common_header* hdr;
+	struct p2m_mqrecv_payload* payload;
```
Comment (yiying-zhang) [inline managers/processor/message_queue.c:67]: empty line

File: managers/processor/message_queue.c [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+	len_msg = sizeof(*hdr)+sizeof(*payload);
```
Comment (yiying-zhang) [inline managers/processor/message_queue.c]: spaces before/after +

File: managers/processor/message_queue.c [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+	payload=to_payload(msg);
+
+
+	copy_from_user(payload->mq_name, name, name_size+1);
+
+	retlen = ibapi_send_reply_imm(current_pgcache_home_node(), msg, len_msg, &retval, sizeof(retval),false);	
+
+	/* check return value */
+	if(retlen == -ETIMEDOUT){
+		return -1;
```
Comment (yiying-zhang) [inline managers/processor/message_queue.c:86]: free msg (no matter what you return

File: managers/processor/message_queue.c [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+	}		
+
+
+
+	/* reply, reply 0 means good */	
+	if(retval.ret == 0){
```
Comment (yiying-zhang) [inline managers/processor/message_queue.c]: use a macro like SUCCESS or MSG_RET_SUCCESS instead of 0

File: managers/processor/message_queue.c [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+
+}
+
+SYSCALL_DEFINE2(mq_close, char*, name, unsigned long, name_size)
+{	
+	ssize_t retval, retlen;
+	u32 len_msg;
+	void *msg;
+	struct common_header* hdr;
+	struct p2m_mqclose_payload* payload;
```
Comment (yiying-zhang) [inline managers/processor/message_queue.c:115]: empty line

File: managers/processor/message_queue.c [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+	payload=to_payload(msg);
+
+	copy_from_user(payload->mq_name, name, name_size+1);
+
+	retlen = ibapi_send_reply_imm(current_pgcache_home_node(), msg, len_msg, &retval, sizeof(retval),false);	
+
+	/* check return value
+    	*/
+	if(retlen == -ETIMEDOUT){
+		return -1;
```
Comment (yiying-zhang) [inline managers/processor/message_queue.c:135]: free msg

File: managers/processor/message_queue.c [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+
+}
+
+SYSCALL_DEFINE3(mq_open, char* , name, unsigned long, name_size, unsigned long, msg_size)
+{
+	ssize_t retval, retlen;
+	u32 len_msg;
+	void *msg;
+	struct common_header* hdr;
+	struct p2m_mqopen_payload* payload;
```
Comment (yiying-zhang) [inline managers/processor/message_queue.c:151]: empty line

File: managers/processor/message_queue.c [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+	hdr->src_nid = LEGO_LOCAL_NID;
+	payload=to_payload(msg);
+
+	copy_from_user(payload->mq_name, name, name_size+1);
+	payload->msg_size=msg_size;
+
+	retlen = ibapi_send_reply_imm(current_pgcache_home_node(), msg, len_msg, &retval, sizeof(retval),false);	
+
+	if(retlen == -ETIMEDOUT){
+		return -1;
```
Comment (yiying-zhang) [inline managers/processor/message_queue.c:170]: free msg

**Review by yiying-zhang — CHANGES_REQUESTED** [review]

File: managers/memory/missing_syscalls.c (hunk 3 above)
Comment (LiuHaolan) [reply managers/memory/missing_syscalls.c:306]: as you don't have sys calls in mComponents, this one is set just to compile successfully in mComponents. Just a hack in LegoOS.

**Pull request merged**